### PR TITLE
feature: add environment and repository file

### PR DESCRIPTION
**Pull request by BrendanThompson**
This PR contains the code to implement `gitlab_resource_file` and `gitlab_environment`, as per the below checklist a few items are still required before this is good to be merged.

One thing that I would like some assistance with however is the fact that sometimes when using the `gitlab_resource_file` resource type the GitLab API returns a `400 Bad Request` when deleting files as well as creating. This is why I have added a sleep into both create and delete functions. It seems to only occur when a few files are being added or deleted. Any advice here would be greatly appreciated.

# ToDo
- [x] Add tests for `gitlab_resource_file`
- [x] Add tests for `gitlab_environment`
- [x] Update website data for the two new resources

### CODE REVIEW

**Comment by BrendanThompson** [comment]
@roidelapluie @ringods @jorcau - would really appreciate any insight you can provide by comment in the above about the random `400 Bad Request` responses.

**Comment by mmurygin** [comment]
Hi @BrendanThompson,

I would like to have repository_file resource merged in master. Can I just fork it and finish the remaining tasks or it's better to wait until you complete them?

**Comment by BrendanThompson** [comment]
@mmurygin I don't believe that there is really anything remaining to be honest. I'm just waiting for some feedback from the maintainers but they seem to be absent at the moment. I can reach out to some Hashicorp people I know. 

Is there a particular feature you feel I've not yet catered for?

**Comment by mmurygin** [comment]
@BrendanThompson,

It looks good to me, and we need to complete tests for file resource and website docs.

As for 400 error, it could be really annoying. May be it worth to ask someone from GitLab?

**Comment by BrendanThompson** [comment]
@mmurygin It looks as though it may have been a transitory thing. I'm going to remove all the sleeps in there. The tests are mostly written too, just haven't pushed them. Have not updated the website stuff at all though. I'll try remember to do it tomorrow 😃

**Comment by mmurygin** [comment]
Hi @BrendanThompson, how is it going? 

Did you have a time to submit your changes?

**Comment by BrendanThompson** [comment]
@mmurygin hey mate, I will push up some changes now.

**Comment by BrendanThompson** [comment]
@mmurygin I have pushed the latest changes, and added tests for `gitlab_repository_file`, happy for you to review this if you'd like. Seems as though the maintainers have gone silent.

File: gitlab/resource_gitlab_repository_file.go
Comment (mmurygin) [inline gitlab/resource_gitlab_repository_file.go:128]: doesn't work when we change file name

Comment (mmurygin) [reply gitlab/resource_gitlab_repository_file.go:128]: I mean it will try to update file which does not exist

Comment (BrendanThompson) [reply gitlab/resource_gitlab_repository_file.go:128]: OK, so whats the exact scenario. Are you updating the `filename` via the `gitlab_repository_file` resource?

Comment (mmurygin) [reply gitlab/resource_gitlab_repository_file.go:128]: file resource for the first apply
```hcl
resource "gitlab_repository_file" "this" {
  project        = gitlab_project.this.id
  file           = "XXXXX.tf"
  content        = filebase64("./XXXXX.tf")
  branch         = "master"
  author_name    = "Terraform"
  author_email   = "terraform@example.com"
  commit_message = "feature: add meow file"
}
```

file resource for the second apply:
```hcl
resource "gitlab_repository_file" "this" {
  project        = gitlab_project.this.id
  file           = "YYYYYY.tf"
  content        = filebase64("./XXXXX.tf")
  branch         = "master"
  author_name    = "Terraform"
  author_email   = "terraform@example.com"
  commit_message = "feature: add meow file"
}
```

here I just changed the filename

Comment (mmurygin) [reply gitlab/resource_gitlab_repository_file.go:128]: btw, tested file create \ update \ delete, and everything works except the above update

so thanks a lot for your work

Comment (BrendanThompson) [reply gitlab/resource_gitlab_repository_file.go:128]: @mmurygin Awesome! Yeah looks like I didn't add that as an attribute to check for change properly in an update scenario. I wrote some stuff to fix it last night but with the change in TF13 I could not get TF to use my local binary 🤦. I will look into it again tonight and see if I can get that fixed.

File: gitlab/resource_gitlab_environment.go
Comment (armsnyder) [inline gitlab/resource_gitlab_environment.go:63]: To help future-proof this resource, can you change your resource ID to include all the bits needed to do a read? Maybe `project_id:environment_id`? This way someone can come in and write an `Import` function later on.

File: gitlab/resource_gitlab_environment.go
Comment (armsnyder) [inline gitlab/resource_gitlab_environment.go:124]: Can you include the original error in these two error format strings?

e.g.

```go
fmt.Errorf("Environment %q failed to stop prior to delete with status %q: %w", d.Id(), stopResp.Status, err)
```

File: gitlab/resource_gitlab_environment_test.go
Comment (armsnyder) [inline gitlab/resource_gitlab_environment_test.go:95]: Ideally your destroy test would be testing the behavior of your destroy function, i.e. that the environment was deleted but the project was not deleted.

Leaving this here as a mental note for #205 and for your own edification, but don't feel obligated to do this now.

File: gitlab/resource_gitlab_environment.go
Comment (armsnyder) [reply gitlab/resource_gitlab_environment.go:63]: Could you make this change to the repository file resource too?

File: gitlab/resource_gitlab_repository_file.go
Comment (armsnyder) [inline gitlab/resource_gitlab_repository_file.go:22]: Can you change this attribute name to `file_path` so that it [matches the API](https://www.terraform.io/docs/extend/hashicorp-provider-design-principles.html#resource-and-attribute-schema-should-closely-match-the-underlying-api)?

File: gitlab/resource_gitlab_repository_file.go
Comment (armsnyder) [inline gitlab/resource_gitlab_repository_file.go:61]: Based on my reading of the [API docs](https://docs.gitlab.com/ee/api/repository_files.html#create-new-file-in-repository) the encoding defaults to text, meaning this resource has a different default than the API. Could you add an `encoding` attribute and default it to `text`?

File: gitlab/resource_gitlab_repository_file.go
Comment (armsnyder) [inline gitlab/resource_gitlab_repository_file.go:106]: Were you going to do this as part of this PR? Just wondering.

File: gitlab/resource_gitlab_repository_file.go
Comment (armsnyder) [inline gitlab/resource_gitlab_repository_file.go:109]: I think you'll always want to add these attributes, even if they haven't changed. `branch` in particular because it's required, but the others as well. Maybe for the others you can use `GetOk` instead of `HasChange`?

File: gitlab/resource_gitlab_environment.go
Comment (armsnyder) [reply gitlab/resource_gitlab_environment.go:124]: Looks like there are some other places where you could be wrapping the error too, like in the delete function for the repository file.

File: website/docs/r/environment.html.markdown
Comment (armsnyder) [inline website/docs/r/environment.html.markdown:14]: You can remove the group from the examples for both resources.

File: gitlab/resource_gitlab_repository_file_test.go
Comment (armsnyder) [inline gitlab/resource_gitlab_repository_file_test.go:21]: Can you add another test step that changes the filename and content?

File: gitlab/resource_gitlab_repository_file.go
Comment (armsnyder) [inline gitlab/resource_gitlab_repository_file.go:18]: `project` and `branch` attributes should both have `ForceNew: true` since they are primary lookup keys of the resource.

**Review by armsnyder — CHANGES_REQUESTED** [review]
Thanks for the contribution! Some comments for ya.

File: gitlab/resource_gitlab_environment.go
Comment (armsnyder) [reply gitlab/resource_gitlab_environment.go:63]: Could you make this change to the repository file resource too?

Comment (BrendanThompson) [reply gitlab/resource_gitlab_environment.go:67]: This has been actioned for environment in: cbf35b79 I am not sure how I am going to address `file` just yet, I want to think on that a bit. I will resolve this though.

File: gitlab/resource_gitlab_environment.go
Comment (armsnyder) [inline gitlab/resource_gitlab_environment.go:62]: Can you change this to `%s:%d` so that the ID format matches the other IDs in this provider?

**Comment by mmurygin** [comment]
Hi @BrendanThompson @armsnyder,

The PR is stuck for a while, I'm ready to jump and finish it, if it's OK? 

or may be @BrendanThompson going to finish it youself?

**Comment by corbanCloud** [comment]
@armsnyder / @mmurygin / @BrendanThompson 

Hey I stumbled across this PR as this is a feature my team at work is looking for and we'd love to chip in to get it across the finish line.

I'm happy to jump in and address the changes requested here but don't want to step on anyone's toes.

My plan is to fork Brendan's fork and address the changes there and open a new PR, that way all the attribution for his work should remain, we would loose the thread here, but could obviously link to it in the new PRs thread for posterity.

Thoughts, comments, concerns?

**Comment by BrendanThompson** [comment]
Hey @corbanCloud thanks mate! Apologies all been a bit off grid. Give me a couple of days to try get this out, trying to fit it in with other stuff has been a pain. If I've not been able to get it done by Monday it's unlikely I'll be able to any time soon, at which point happy for you to takeover the PR.

**Comment by corbanCloud** [comment]
Hey no sweat at all.
I am more excited to see someone already started than anything! If nothing else thanks for getting this started.

I'll keep an eye on the thread, thanks for the update.

File: gitlab/resource_gitlab_repository_file.go
Comment (BrendanThompson) [reply gitlab/resource_gitlab_repository_file.go:18]: I have added this to `file` too.

File: website/docs/r/environment.html.markdown
Comment (BrendanThompson) [reply website/docs/r/environment.html.markdown:14]: No, this ensures that the example is working.

File: gitlab/resource_gitlab_repository_file.go
Comment (BrendanThompson) [reply gitlab/resource_gitlab_repository_file.go:61]: When I initially started working on this I did have the encoding as something that could be set, however it derped when you passed in `text`

**Comment by nicholasklick** [comment]
@corbanCloud @mmurygin are either of you interested in jumping in to help finish this?

**Comment by corbanCloud** [comment]
Hey @nicholasklick I'm still keen to jump in, looks like Brendan isn't able to finish based on his own timelines. I can work on it a bit this weekend, if you're free before then, by all means take it away

**Comment by BrendanThompson** [comment]
@corbanCloud I knocked off a few things the other day, just need to deal with the tests etc. Do you want me to add you as a collaborator to my fork, might make it easier?

**Comment by corbanCloud** [comment]
Sure! I thought you had mentioned something about not being able to get to it at all if you didn't get to it by last Monday or something

**Comment by BrendanThompson** [comment]
@corbanCloud yeah I managed to sneak in 30mins last week hahah

**Comment by nicholasklick** [comment]
@corbanCloud @mmurygin just following up. Do you need any assistance with wrapping this up?

**Comment by corbanCloud** [comment]
Hey @nicholasklick  

I haven't been able to make personal time for this as I hoped, but I have moved this forward within my day job as we've got a growing need for this capability.

We've got a ticket in our current workstream to chip in to wrap this up so either myself or one of our engineers will pick up the work, and you should have a PR in the near future. One or two weeks maybe, I'll comment back here if that timeline changes.

**Comment by BrendanThompson** [comment]
@corbanCloud – added you as a contributor awhile ago for my fork if you wanted to continue it there. I have closed out most of the issues that were raised by @armsnyder , I think its just tests left.

**Comment by corbanCloud** [comment]
@BrendanThompson Can you resend? my invite expired

**Comment by BrendanThompson** [comment]
@corbanCloud – I have done that now mate 😃.

**Comment by mehmettaskiner** [comment]
It would be great to have this feature merged, I am not a go expert but glad to help if you need anything.

**Comment by garrett-EY** [comment]
Hi @BrendanThompson, I'm on @corbanCloud's team and can help work on this. Could you please add me as a contributor as well?

**Comment by BrendanThompson** [comment]
@garrett-EY I have just done this now 😃. Thanks Garrett!

**Comment by garrett-EY** [comment]
@corbanCloud and I are starting to work on this, we'll keep you all posted. 👍

File: gitlab/resource_gitlab_repository_file.go
Comment (garrett-EY) [reply gitlab/resource_gitlab_repository_file.go:64]: Yep I can confirm this, passing in "text" to `encoding` returns a 400. It also seems that all file content is base64 encoded even if you pass plaintext as `content` and emit `encoding` entirely.

**Comment by springroll12** [comment]
Any updates on this?

**Comment by garrett-EY** [comment]
@BrendanThompson @corbanCloud I've started work on this in a branch off your fork, but I've been running into some interesting issues. Could you weigh in when you get a chance? Thanks!

https://github.com/BrendanThompson/terraform-provider-gitlab/pull/1

File: gitlab/resource_gitlab_repository_file.go
Comment (corbanCloud) [reply gitlab/resource_gitlab_repository_file.go:64]: @armsnyder Can you offer any clarity here it seems there some misaligned expectations on what we should do as the request to default to `text` encoding seems to cause a 400 error as base64 seems to be the only valid encoding?

Comment (corbanCloud) [reply gitlab/resource_gitlab_repository_file.go:64]: @nicholasklick probably should have cc'd you in the above comment as well. Looks like use of anything other than base64 for encoding is causing issues?

**Comment by corbanCloud** [comment]
@nicholasklick @armsnyder We've got some issues with request encoding, would love some input. Pinged you in the conversation above but havent heard back, no hard feelings on my end I know everyone is busy https://github.com/gitlabhq/terraform-provider-gitlab/pull/350#pullrequestreview-487253837

In the interim, @garrett-EY and I are going to push ahead w/ the encoding fixed to base64 as the text encoding is causing issues for both us and @BrendanThompson. Our team has some more availability to contribute to this coming up and would like to try and complete this or get as close as possible and revisit the encoding discussion.

**Comment by BrendanThompson** [comment]
@corbanCloud yeah I was never able to get it working without enforcing the encoding unfortunately!

**Comment by corbanCloud** [comment]
@armsnyder @nicholasklick 
Looking for input here, can we move ahead with this PR w/o the requested change to encoding? None of us have been able to make `text` encoding work to date

https://github.com/gitlabhq/terraform-provider-gitlab/pull/350#discussion_r487445170

**Comment by armsnyder** [comment]
@corbanCloud Sorry for the delayed reply. I did some tests and came to the conclusion that the only allowed values for `encoding` are `nil` and `"base64"`. Since setting it to `"base64"` causes it to behave differently, it should not be the default. Can you leave the Encoding field unset if the `encoding` attribute is `"text"` as a workaround for this API bug, with a good comment?

### Some example API calls to illustrate
#### Default encoding
```
➜  curl -H 'PRIVATE-TOKEN: ACCTEST' 'http://localhost:8080/api/v4/projects/74/repository/files/baz.txt' \
  -F content=hello -F branch=master -F commit_message=msg
{"file_path":"baz.txt","branch":"master"}
```
```
➜  ~ curl -H 'PRIVATE-TOKEN: ACCTEST' 'http://localhost:8080/api/v4/projects/74/repository/files/baz.txt/raw?ref=master'
hello
```
#### Specify `base64` encoding
```
➜  curl -H 'PRIVATE-TOKEN: ACCTEST' 'http://localhost:8080/api/v4/projects/74/repository/files/baz2.txt' \
  -F content=hello -F branch=master -F commit_message=msg -F encoding=base64
{"file_path":"baz2.txt","branch":"master"}
```
```
➜  ~ curl -H 'PRIVATE-TOKEN: ACCTEST' 'http://localhost:8080/api/v4/projects/74/repository/files/baz2.txt/raw?ref=master'
??e
```
#### Specify `text` encoding
```
➜  curl -H 'PRIVATE-TOKEN: ACCTEST' 'http://localhost:8080/api/v4/projects/74/repository/files/baz2.txt' \
  -F content=hello -F branch=master -F commit_message=msg -F encoding=text
encoding does not have a valid value
```

**Comment by corbanCloud** [comment]
@armsnyder No sweat, thanks for your testing, that reflects our findings as well. 
I'll work w/ @garrett-EY to get that going

**Comment by ckoliber** [comment]
When this feature will be available ?

**Comment by timofurrer** [comment]
The `gitlab_repository_file` resource will be added with #724 . Can you please create a new PR with the `gitlab_environment` resource. I'm going to close this for now - since it anyways needs quite some refactoring after the SDK v2 upgrade which has been landed.

**Pull request closed without merging**
